protovalidate/internal/extra_func.py

```diff
@@ -161,28 +170,50 @@ def is_email(string: celtypes.Value) -> celpy.Result:
 
 
 def is_uri(string: celtypes.Value) -> celpy.Result:
-    url = urlparse.urlparse(str(string))
-    # urlparse correctly reads the scheme from URNs but parses everything
-    # after (except the query string) as the path.
-    if url.scheme == "urn":
-        if not (url.path):
-            return celtypes.BoolType(False)
-    elif not all([url.scheme, url.netloc, url.path]):
-        return celtypes.BoolType(False)
+    """Validate whether string is a valid URI.
+
+    URI is defined in the internet standard RFC 3986.
+    Zone Identifiers in IPv6 address literals are supported (RFC 6874).
+
+    Args:
+        string (celTypes.Value): The string to validate.
 
-    # If the query string contains percent-encoding, then try to decode it.
-    # unquote will return the same string if it is improperly encoded.
-    if "%" in url.query:
-        return celtypes.BoolType(urlparse.unquote(url.query) != url.query)
+    Returns:
+        True if the string is a URI, for example "https://example.com/foo/bar?baz=quux#frag". False otherwise.
 
-    return celtypes.BoolType(True)
+    Raises:
+        celpy.CELEvalError: If string is not an instance of celtypes.StringType.
```

Comment:
Is it typical in Python to document arguments, return value, and exceptions like this?

Reply:
Yeah, from the [PEP proposal](https://peps.python.org/pep-0257/#multi-line-docstrings) you linked:

> The docstring for a function or method should summarize its behavior and document its arguments, return value(s), side effects, exceptions raised, and restrictions on when it can be called (all if applicable). Optional arguments should be indicated. It should be documented whether keyword arguments are part of the interface.

There are a few styles for doing this. This one uses the [Google style guide](https://google.github.io/styleguide/pyguide.html#383-functions-and-methods), which is also used in Effective Python.

Reply:
I see we're also using a similar style in validator.py.

We definitely want to follow best practices, but there's some leeway. Don't need to completely rewrite the docs for every port.

Reply:
Yep, fair enough. Will keep that in mind going forward.